### PR TITLE
fix: keep HEAD_DIM 256 prefill within 64 KiB smem on SM75/SM70

### LINKED ISSUE #3528: head_dim 256 prefill fails to launch with cudaErrorInvalidValue on SM75/SM70 (SharedStorageQKVO smem overflow)
### Summary

`HEAD_DIM 256` prefill kernels fail to launch on Turing/Volta (SM75/SM70) with
`cudaErrorInvalidValue` ("invalid argument"). The prefill `SharedStorageQKVO` struct exceeds the
64 KiB opt-in shared-memory cap (`cudaDevAttrMaxSharedMemoryPerBlockOptin`) on these arches, so
`cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, ...)` is rejected and
the kernel never runs.

### Root cause

`SharedStorageQKVO` unconditionally reserved 1-element placeholder arrays for the NVFP4
scale-factor buffers (`k_sf_smem` / `v_sf_smem`) and the FP8 "repack" buffer (`kv_smem_repack`)
even for KV dtypes that never touch them. After `alignas(16)` each placeholder occupies 16 B,
inflating `sizeof(SharedStorageQKVO)` by 48 B (FP16) / 32 B (FP8).

For `HEAD_DIM 256` the q/k/v union already fills the entire 64 KiB opt-in budget, so the extra
padding pushes the struct past 65536 B:

| KV dtype | `sizeof(SharedStorageQKVO)` @ HEAD_DIM 256 | 64 KiB cap |
| -------- | ------------------------------------------ | ---------- |
| FP16     | 65584                                       | exceeds    |
| FP8      | 65568                                       | exceeds    |
| NVFP4    | 50192                                       | fits       |

On SM80+ the opt-in cap is larger (≥ 100 KiB), so the overflow is invisible there and not
exercised by upstream CI; the failure is specific to SM75/SM70.

### Reproduction

- Hardware: NVIDIA RTX 2080 Ti (`sm_75`)
- Any `head_dim = 256` FP16 or FP8-KV prefill, e.g. `single_prefill_with_kv_cache` /
  `BatchPrefillWithPagedKVCache` (originally observed via vLLM `BatchPrefillWithPagedKVCache`).

```python
import torch, flashinfer
S, Skv, H, D = 128, 512, 4, 256  # head_dim 256 -> SM75 smem-overflow path
q = torch.randn(S, H, D, device="cuda", dtype=torch.float16)
k = torch.randn(Skv, H, D, device="cuda", dtype=torch.float16)
v = torch.randn(Skv, H, D, device="cuda", dtype=torch.float16)
o = flashinfer.single_prefill_with_kv_cache(q, k, v)  # launch fails: invalid argument on sm_75
```

Observed error:

```
Check failed: (status == cudaSuccess) ... kernel launch failed, error: invalid argument
```

### Expected behavior

`head_dim 256` prefill should launch and produce correct results on SM75/SM70, as it does on
SM80+.

### Proposed fix

Move the conditional buffers into empty-base mixins so that, for dtypes that don't use them, they
contribute 0 bytes via the empty base optimization instead of a padded placeholder. This brings
FP16/FP8 back to exactly 65536 B at `HEAD_DIM 256` (NVFP4 unchanged), with no behavior change for
dtypes that do use the buffers and no change on SM90+.

PR: #3526

### Environment

- FlashInfer: `main`
- GPU: RTX 2080 Ti (`sm_75`, Turing); also affects SM70 (Volta)
- CUDA: 12.8

### PR DESCRIPTION
**Pull request by HuChundong**
## 📌 Description

`SharedStorageQKVO` unconditionally reserved 1-element placeholder arrays for the NVFP4
scale-factor buffers (`k_sf_smem` / `v_sf_smem`) and the FP8 "repack" buffer
(`kv_smem_repack`) even for KV dtypes that never touch them. After `alignas(16)` each
placeholder occupies 16 B, inflating `sizeof(SharedStorageQKVO)` by 48 B (FP16) / 32 B (FP8).

For `HEAD_DIM 256` the q/k/v union already fills the entire 64 KiB opt-in shared memory budget
(`cudaDevAttrMaxSharedMemoryPerBlockOptin`) on Turing/Volta (SM75/SM70). The extra padding
pushes the struct past 65536 B, so
`cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, ...)` is rejected
with `cudaErrorInvalidValue` ("invalid argument") and the prefill kernel never launches. This
is a regression for `head_dim 256` on these arches (e.g. observed via vLLM
`BatchPrefillWithPagedKVCache` on an RTX 2080 Ti).

**Fix:** move the conditional buffers into empty-base mixins (`KVScaleFactorSmem`,
`KVRepackSmem`) so that, for dtypes that don't use them, they contribute 0 bytes via the
empty base optimization instead of a padded placeholder. They are accessed through accessors
(`k_sf_smem_ptr` / `v_sf_smem_ptr` / `kv_smem_repack_ptr`) that return `nullptr` when the
buffer is absent; the result is only ever dereferenced inside the matching `if constexpr`
branch, and the offset is folded into the accessor so it is never applied to `nullptr`. No
behavior change for dtypes that do use the buffers (NVFP4 scale factors, FP8 repack), and no
change on SM90+.

`sizeof(SharedStorageQKVO)`, `HEAD_DIM 256` (64 KiB cap):

| KV dtype | before | after  | result                          |
| -------- | ------ | ------ | ------------------------------- |
| FP16     | 65584  | 65536  | now fits                        |
| FP8      | 65568  | 65536  | now fits                        |
| NVFP4    | 50192  | 50176  | already fit; unchanged behavior |

## 🔍 Related Issues

Fixes #3528 — the `head_dim 256` prefill `cudaErrorInvalidValue` / "invalid argument" launch
failure on SM75/SM70 (Turing/Volta). Reproduced on an RTX 2080 Ti (sm_75) via vLLM
`BatchPrefillWithPagedKVCache`.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks on the changed files and fixed any reported issues (clang-format
      `v19.1.1` and all other configured hooks pass; no reformatting required).

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed — no new tests required; `head_dim 256`
      functional coverage already exists in the prefill suite and the change is an internal
      smem-layout fix.
- [x] Relevant tests pass on affected hardware (see below).

This is a shared-memory sizing fix; functional coverage for `head_dim 256` already exists in
the prefill test suite (`tests/attention/test_single_prefill.py`,
`tests/attention/test_batch_prefill_kernels.py`, `tests/attention/test_fp8_prefill.py`, which
parametrize `head_dim=256`). The bug only manifests on SM75/SM70, whose 64 KiB opt-in
shared-memory cap is not exercised by upstream CI (SM80+).

**Test results on an RTX 2080 Ti (sm_75)** with this fix (these all fail to launch *without*
the fix — `cudaErrorInvalidValue`):

| Suite (`-k "256"`)                                    | Result                      |
| ----------------------------------------------------- | --------------------------- |
| `test_single_prefill.py` (FP16)                       | 60 passed                   |
| `test_batch_prefill_kernels.py` (FP16, paged/ragged)  | 868 passed                  |
| `test_fp8_prefill.py` (FP8 KV)                         | 224 passed                  |

Note: **NVFP4** `head_dim 256` cases are skipped/expected-fail on this arch — Turing has no
FP4 tensor cores, unrelated to this change (NVFP4 footprint is unchanged anyway).

Additional validation:

- **Footprint:** `sizeof(SharedStorageQKVO)` for `HEAD_DIM 256` drops to ≤ 65536 B for FP16
  and FP8 (see table above); NVFP4 unchanged.
- **Compile:** the FA2 single / batch-paged / batch-ragged prefill kernels
  (`CTA_TILE_Q` 16 / 64 / 128, `HEAD_DIM 256`, FP16 and FP8 KV) compile for `sm_75`.
- **Hardware (RTX 2080 Ti, sm_75):** `cudaFuncSetAttribute(...MaxDynamicSharedMemorySize...)`
  accepts 65536 B and the kernel launches, while 65584 B is rejected with
  `cudaErrorInvalidValue` — exactly the before/after boundary of this fix.
- **Numerical:** `head_dim=256` FP16 `single_prefill_with_kv_cache` matches a PyTorch
  reference (`allclose`, max abs diff ≈ 2.4e-4).

## Reviewer Notes

- The change is gated entirely by KV dtype via `if constexpr`; the accessors return `nullptr`
  only on paths that are never taken for those dtypes, so there is no extra runtime branch in
  the active code paths.
- Empty base optimization (rather than a zero-length array) keeps this strictly
  standard-conforming C++17 while still contributing 0 bytes.
- No changes to public APIs or kernel behavior; only the internal smem struct layout for
  dtypes that previously over-reserved.

## Summary by CodeRabbit

* **Refactor**
  * Internal rework of shared-memory handling in attention routines (FP4/FP8 paths) with new accessor-based storage layout.
  * Updated attention kernel implementations to use the new storage accessors.
  * No changes to public APIs or user-visible behavior; maintenance and reliability improvements under the hood.

### CODE REVIEW

**Comment by Zephyr69** [comment]
This fix works! Could you adapt it to newer versions of flashinfer?

**Review by saltyminty — APPROVED** [review]
Approved conditional on CI

**Comment by saltyminty** [comment]
Pipeline looks good.

**Comment by ir1ka** [comment]
Looking forward to the merge, but the pre-commit check failed.
